src/backend/distributed/executor/multi_utility.c

```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+		char *escapedCommandString = NULL;
+		StringInfo applyCommand = makeStringInfo();
+
+		shardConnections = GetShardConnections(shardConnectionHash,
+											   shardId,
+											   &shardConnectionsFound);
+		Assert(shardConnectionsFound);
 
 		/* build the shard ddl command */
-		char *escapedCommandString = quote_literal_cstr(commandString);
-		StringInfo applyCommand = makeStringInfo();
+		escapedCommandString = quote_literal_cstr(commandString);
 		appendStringInfo(applyCommand, WORKER_APPLY_SHARD_DDL_COMMAND, shardId,
 						 escapedCommandString);
 
-		shardPlacementList = FinalizedShardPlacementList(shardId);
-		foreach(shardPlacementCell, shardPlacementList)
-		{
-			ShardPlacement *placement = (ShardPlacement *) lfirst(shardPlacementCell);
-			char *workerName = placement->nodeName;
-			uint32 workerPort = placement->nodePort;
+		ExecuteCommandOnShardReplicas(applyCommand, shardId, shardConnections);
 
-			List *queryResultList = ExecuteRemoteQuery(workerName, workerPort,
-													   relationOwner, applyCommand);
-			if (queryResultList == NIL)
-			{
-				/*
-				 * If we failed on the first placement, return false. We return
-				 * here instead of exiting at the end to avoid breaking through
-				 * multiple loops.
-				 */
-				if (isFirstPlacement)
-				{
-					return false;
-				}
+		FreeStringInfo(applyCommand);
+	}
 
-				ereport(WARNING, (errmsg("could not apply command on shard "
-										 UINT64_FORMAT " on node %s:%d", shardId,
-										 workerName, workerPort),
-								  errdetail("Shard placement will be marked as "
-											"inactive.")));
+	/* check for cancellation one last time before returning */
+	CHECK_FOR_INTERRUPTS();
 
-				*failedPlacementList = lappend(*failedPlacementList, placement);
-			}
-			else
-			{
-				ereport(DEBUG2, (errmsg("applied command on shard " UINT64_FORMAT
-										" on node %s:%d", shardId, workerName,
-										workerPort)));
-			}
+	/* CloseConnections(allShardsConnectionList); */
 
-			isFirstPlacement = false;
-		}
+	return true;
+}
 
-		FreeStringInfo(applyCommand);
+
+/*
+ * OpenConnectionsToAllShardPlacement opens connections to all placements of
+ * the given shard Id Pointer List and returns the hash table containing the connections.
+ * The resulting hash table maps shardIds to ShardConnection structs.
+ */
+static HTAB *
+OpenConnectionsToAllShardPlacements(List *shardIdList, char *relationOwner)
+{
+	HTAB *shardConnectionHash = CreateShardConnectionHash();
+	ListCell *shardCell = NULL;
+	MemoryContext oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	foreach(shardCell, shardIdList)
+	{
+		uint64 *shardIdPointer = (uint64 *) lfirst(shardCell);
+		uint64 shardId = (*shardIdPointer);
+
+		OpenConnectionsToShardPlacements(shardId, shardConnectionHash, relationOwner);
 	}
 
-	return true;
+	shardPlacementConnectionList = ConnectionList(shardConnectionHash);
+
+	MemoryContextSwitchTo(oldContext);
+
+	if (!isXactCallbackRegistered)
```

Comment:
For integration purposes, maybe move this if-block into an EnableXactCallback function (or better name).

e.g.
https://github.com/citusdata/citus-mx/blob/4c70aba5b300be28bb3a81187591246ef16b390f/src/backend/distributed/transaction/worker_transaction.c#L200